src/DynamoCoreWpf/ViewModels/Watch3D/DynamoGeometryModel3D.cs

```diff
@@ -13,24 +14,187 @@
+                    + (int)(HasTransparencyData ? DynamoMeshShaderFlags.HasTransparency : 0)
+                     + (int)(RequiresPerVertexColor ? DynamoMeshShaderFlags.RequiresPerVertexColor : 0)
+                      + (int)(IsFlatShaded ? DynamoMeshShaderFlags.FlatShade : 0);
+
+            return finalFlag;
+        }
+
+        protected override void OnUpdatePerModelStruct(RenderContext context)
+        {
+            base.OnUpdatePerModelStruct(context);
+            //store the entire state of all our flags in the X component and decode in the vertex shader.
+            //we'll pass this from the vertex shader to the fragment shader to determine color states.
+            //Params is a helix material builtin and maps to vParams in the shader.
+
+            //TODO this gets called whenever a render occurs, make sure it is fast.
+            modelStruct.Params.X = GenerateEnumFromState();
+        }
+
+        internal void SetPropertyData(DependencyPropertyChangedEventArgs args )
+        {
+            var depType = args.Property;
+            var argval = (bool)args.NewValue;
+            //use dependencyProperty to determine which data to set.
+            if (depType == AttachedProperties.ShowSelectedProperty )
+            {
+                IsSelectedData = argval;
+            }
+
+            else if (depType == AttachedProperties.IsFrozenProperty)
+            {
+                IsFrozenData = argval;
+            }
+
+            else if (depType == AttachedProperties.IsolationModeProperty)
+            {
+                IsIsolatedData = argval;
+            }
+
+            else if (depType == AttachedProperties.IsSpecialRenderPackageProperty)
+            {
+                IsSpecialRenderPackageData = argval;
+            }
+
+            else if (depType == AttachedProperties.HasTransparencyProperty)
+            {
+                HasTransparencyData = argval;
+            }
+
+            else if (depType == DynamoGeometryModel3D.RequiresPerVertexColorationProperty)
+            {
+                RequiresPerVertexColor = argval;
+            }
+
+
+            //TODO we need to add FlatShader to AttachedProperties if we want to use it.
+            //and add a case here.
+        }
+    }
+
+
     /// <summary>
-    /// A Dynamo mesh class which supports the RenderCustom technique.
+    /// A Dynamo mesh class which supports sending data to our custom shader.
     /// </summary>
-    public class DynamoGeometryModel3D : MeshGeometryModel3D
+    public class DynamoGeometryModel3D : MaterialGeometryModel3D
     {
         public DynamoGeometryModel3D()
         {
         }
 
         protected override SceneNode OnCreateSceneNode()
         {
-            var node = base.OnCreateSceneNode();
-            node.OnSetRenderTechnique = (host) => { return host.EffectsManager["RenderCustom"]; };
+            var node = new DynamoMeshNode();
             return node;
```

Comment:
`return new DynamoMeshNode();`

Reply:
So by overriding this method, helix knows which custom core and custom shaders to use to render Dynamo mesh `GeometryModel3D` types?

Reply:
only to render `DynamoGeometryModel3d` - not GeometryModel3D in general.

Reply:
Yeah, that's what I meant. Thanks.